Fix: align behaviour to tsc `rewriteRelativeImportExtensions`

JLHwung opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/17114
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/3052
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->
In this PR we introduced a new helper `tsRewriteRelativeImportExtensions` that is almost identical (the regex is slightly improved with the help of eslint-regex) to the TS helper. The new behaviour in this PR should mirror the tsc `rewriteRelativeImportExtensions` option without the tsc JSX output option.

nicolo-ribaudo reviewed (commented):
Would just tweaking our original regexp to this have the same effect?
```js
.replace(/^(\.\.?[\\/].*\.[mc]?)tsx?$/, "$1js")
```

JLHwung commented:
> Would just tweaking our original regexp to this have the same effect?
> 
> ```js
> .replace(/^(\.\.?[\\/].*\.[mc]?)tsx?$/, "$1js")
> ```

Good question. I think this one still deviates from the TS one: 1) It does not handle `.d.ts` and 2) It does not support case-insensitivie variants such as `.mTs`. Though the latter is trivial to fix, the former still requires longer regex. I added a unit test for the new helper and it turns out we have found a new bug in this PR, I will fix that later.

JLHwung commented on packages/babel-helpers/test/unittests/tsRewriteRelativeImportExtensions.test.js:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  it.each([
+    // skip string wrapper object
```
The current Babel transform results will fill these tests because we implicitly applied `toString()` on the input.

JLHwung commented:
Ok, it turns the bug was intentional as tsc want to leave `.d.css.ts` intact. I have reverted to the original implementation and added more test cases. This PR is ready to review.

JLHwung commented:
I also open a docs PR to clearly state that the `rewriteImportExtensions` option mirrors the tsc `rewriteRelativeImportExtensions` option: https://github.com/babel/website/pull/3052

nicolo-ribaudo commented:
I am a bit worried about the effects of this PR when it comes to `from "@some/package/foo.ts`: the current behavior is intentional (https://github.com/babel/babel/pull/15913#discussion_r1313010447) and can be useful for example when working with monorepos and importing specific files from other packages.

Do you think it'd be possible for now to only change the behavior for absolute specifiers, and keep the complete alignment for Babel 8?

JLHwung commented:
The TS team also intentionally does not support `@foo/bar.ts`: https://github.com/microsoft/TypeScript/pull/59767#issuecomment-2394281978. 

For the `@some/package/foo.ts` usage, I think it'd better resort to the extension-less sub-exports like `@some/package/foo`. I just checked our monorepo and fortunately, we don't have such usage. I suggest we ship it in the next minor and gather some feedback. If there is strong pushback, we can defer the scoped package part to Babel 8.

Since our `rewriteImportExtensions` option (Babel 7.23 in Sep 2023) preceded tsc's `rewriteRelativeImportExtensions` option (TS 5.7 in Nov 2024), it can be understood that they are not aligned at this moment. The sooner we close this gap, the less friction the ecosystem will feel.

nicolo-ribaudo commented:
Ok, let's do it.

nicolo-ribaudo commented on packages/babel-preset-typescript/src/plugin-rewrite-ts-imports.ts:
```diff
@@ -1,25 +1,50 @@
+      if (/^\.\.?\//.test(source.value)) {
+        // @see packages/babel-helpers/src/helpers/tsRewriteRelativeImportExtensions.ts
+        source.value = source.value.replace(
+          /\.(tsx)$|((?:\.d)?)((?:\.[^./]+)?)\.([cm]?)ts$/i,
```
> `[^./]`

Should we also include `\` for windows paths?

JLHwung replied:
Good question. TS currently does not handle `\`: https://www.typescriptlang.org/play/?rewriteRelativeImportExtensions=true#code/JYWwDg9gTgLgBAIgHRIDoDMISTAzggKFElkRQHpNs8Eg

I think we should handle the windows path separator as well. I will open a new issue to the TS.

nicolo-ribaudo commented on packages/babel-preset-typescript/src/plugin-rewrite-ts-imports.ts:
```diff
@@ -1,25 +1,50 @@
-        source.value = source.value
-          .replace(/(\.[mc]?)ts$/, "$1js")
-          .replace(/\.tsx$/, ".js");
+      if (/^\.\.?\//.test(source.value)) {
```
```suggestion
      if (source.value.startsWith(".")) {
```

JLHwung replied:
This regex tests about `./` and `../`, so it can not be replaced by the suggestion.

nicolo-ribaudo commented on packages/babel-preset-typescript/src/plugin-rewrite-ts-imports.ts:
```diff
@@ -31,22 +56,28 @@ export default declare(function ({ types: t, template }) {
       },
-      CallExpression(path) {
+      CallExpression(path, state) {
         if (t.isImport(path.node.callee)) {
```
While we are changing this, could you wrap this in a `!process.env.BABEL_8_BREAKING`?

nicolo-ribaudo commented on packages/babel-helpers/src/helpers/tsRewriteRelativeImportExtensions.ts:
```diff
@@ -0,0 +1,23 @@
+  path: unknown,
+  preserveJsx?: boolean,
+) {
+  if (typeof path === "string" && /^\.\.?\//.test(path)) {
```
```suggestion
  if (typeof path === "string" && path[0] === ".") {
```

nicolo-ribaudo replied:
See https://github.com/babel/babel/pull/17118#discussion_r1951443136

nicolo-ribaudo commented:
@JLHwung This can go in a patch release, right?

JLHwung commented:
> @JLHwung This can go in a patch release, right?

I think it's time for a new minor given that 7.26 was released 3 months ago. If so maybe this PR can be shipped with 7.27.

JLHwung commented:
Assuming this PR will be shipped in v7.27.0, I have updated the `minVersion` and cleared the todo item.
